main/streams/streams.c

```diff
@@ -692,12 +692,21 @@ PHPAPI size_t _php_stream_read(php_stream *stream, char *buf, size_t size)
 
 		if (!stream->readfilters.head && (stream->flags & PHP_STREAM_FLAG_NO_BUFFER || stream->chunk_size == 1)) {
 			toread = stream->ops->read(stream, buf, size);
-			if (toread == (size_t) -1) {
-				/* e.g. underlying read(2) returned -1 */
+			if (toread < 0) {
+				/* Report an error if the read failed and we did not read any data
+				 * before that. Otherwise return the data we did read. */
+				if (didread == 0) {
+					return toread;
+				}
 				break;
 			}
 		} else {
-			php_stream_fill_read_buffer(stream, size);
+			if (php_stream_fill_read_buffer(stream, size) != SUCCESS) {
+				if (didread == 0) {
+					return -1;
+				}
+				break;
+			}
 
 			toread = stream->writepos - stream->readpos;
 			if (toread > size) {
```

Comment:
@nikic, this is an example where cl.exe throws a warning, since we're comparing a `ssize_t` with a `size_t` value. If we're sure that `toread >= 0` we could simply apply a cast.

Reply:
I've inserted casts for the two cases that you mentioned, though generally I think that it would be better to suppress those warnings for the MSVC build. We don't have them enabled in the gcc/clang builds.

Reply:
I think these are very valid warnings, and instead of suppresing them, we should "fix" the respective code. In this case, for instance, if `toread` was ever negative, the condition would almost certainly evaluate to true, which is unlikely to be desired. If we're **sure** that `toread` can never be negative, adding a cast clarifies this for the compiler and for readers. (If we're not **absolutely** sure, we could add an additional assertion.)

Reply:
Perhaps I could attract your attention to some macros in Zend/zend_range_check.h. Something like `ZEND_SIZE_T_GT_ZEND_LONG` could suffice for the runtime checks. Could also be wrapped with `UNEXPECTED` where makes sense. Developed back then in 7.0 when there was a lot of stuff to fix for the `size_t` migration :)

Thanks.